packages/ai-jsx/src/core/conversation.tsx

```diff
@@ -0,0 +1,295 @@
+ *      <FunctionResponse name="evaluate_math">83076</FunctionResponse>
+ *    </ChatCompletion>
+ *
+ *    ==> "That would be 83,076."
+ * ```
+ */
+export async function FunctionResponse(
+  { name, children }: { name: string; children: Node },
+  { render }: AI.ComponentContext
+) {
+  const output = await render(children);
+  return `function ${name} returns ${output}`;
+}
+
+interface ConversationMessageType<T, C extends AI.Component<any>> {
+  type: T;
+  element: AI.Element<AI.PropsOfComponent<C>>;
+}
+
+/**
+ * A type that represents a conversation message.
+ */
+export type ConversationMessage =
+  | ConversationMessageType<'user', typeof UserMessage>
+  | ConversationMessageType<'assistant', typeof AssistantMessage>
+  | ConversationMessageType<'system', typeof SystemMessage>
+  | ConversationMessageType<'functionCall', typeof FunctionCall>
+  | ConversationMessageType<'functionResponse', typeof FunctionResponse>;
+
+/** @hidden */
+export function isConversationalComponent(element: AI.Element<any>): boolean {
+  return (
+    element.tag === UserMessage ||
+    element.tag === AssistantMessage ||
+    element.tag === SystemMessage ||
+    element.tag === FunctionCall ||
+    element.tag === FunctionResponse
+  );
+}
+
+function toConversationMessages(partialRendering: AI.PartiallyRendered[]): ConversationMessage[] {
+  return partialRendering
+    .flatMap((e) => (AI.isElement(e) ? [e] : []))
+    .map<ConversationMessage>((e) => {
+      switch (e.tag) {
+        case UserMessage:
+          return { type: 'user', element: e };
+        case AssistantMessage:
+          return { type: 'assistant', element: e };
+        case SystemMessage:
+          return { type: 'system', element: e };
+        case FunctionCall:
+          return { type: 'functionCall', element: e };
+        case FunctionResponse:
+          return { type: 'functionResponse', element: e };
+        default:
+          throw new AIJSXError(
+            `Unexpected tag (${e.tag.name}) in conversation`,
+            ErrorCode.UnexpectedPartialRenderResult,
+            'internal'
+          );
+      }
+    });
+}
+
+/** @hidden */
+export async function renderToConversation(
+  conversation: AI.Node,
+  render: AI.ComponentContext['render']
+): Promise<ConversationMessage[]> {
+  return toConversationMessages(await render(conversation, { stop: isConversationalComponent }));
+}
+
+/**
+ * A component that appends messages to a conversation according to a `reply` function.
+ *
+ * The `reply` function is invoked with the new messages produced from its _own replies_ until
+ * it fails to return a conversational message (e.g. UserMessage or AssistantMessage). The first
+ * invocation includes the messages from the `children` prop.
+ *
+ * @example
+ * ```tsx
+ *    <Converse reply={(messages) => messages[0].type === "user" ? <AssistantMessage>Hello there!</AssistantMessage> : null}>
+ *      <UserMessage>Hello!</UserMessage>
+ *    </ChatCompletion>
```

Comment:
I appreciate the simplicity, but I think this might be a bit of a confusing example, because it seems pretty unlikely that a user would directly render an `AssistantMessage` they generate themselves, right?

Maybe a better example would be driving an automatic user response based on looking at what the previous message was? Like, the pattern where the AI emits a response, and the user responds "are you sure you followed all the instructions?" or "please critique your own response and amend it as necessary".

Reply:
Yeah I was trying to come up with the simplest example that demonstrates the semantics of `<Converse>` which might be more confusing than just showing an example with `<ChatCompletion>`. Will rework this

Reply:
I ended up doing something along the lines of the original motivation, which is user-managed control over the function calling loop. (Which is on its own a little lame because it's roughly what `<UseToolsFunctionCall>` already does. But right now my intuition is that we'll ultimately want to delegate more control of the loop to user code rather than having `<UseTools>` as a one-size-fits-all.)